src/odbc_connection.cpp

```diff
@@ -2786,111 +2930,351 @@ fetch_and_store
+            // Instead, call SQLGetData, and adjust buffer size accordingly
+            if (data->columns[column_index]->is_long_data)
+            {
+              SQLPOINTER target_buffer;
+              SQLLEN     buffer_size =
+                           data->query_options.initial_long_data_buffer_size;
+              SQLLEN     string_length_or_indicator;
+              SQLLEN     data_returned_length = 0;
+
+              if (data->columns[column_index]->bind_type == SQL_C_WCHAR)
+              {
+                row[column_index].wchar_data = (SQLWCHAR *)malloc(buffer_size);
+                // if bad malloc, indicate and return
+                if (row[column_index].wchar_data == NULL)
                 {
-                  row[column_index].size = data->bound_columns[column_index].length_or_indicator_array[row_index];
-                  row[column_index].char_data = new SQLCHAR[row[column_index].size]();
-                  memcpy(
-                    row[column_index].char_data,
-                    (SQLCHAR *)data->bound_columns[column_index].buffer + row_index * data->columns[column_index]->buffer_size,
-                    row[column_index].size
-                  );
-                  break;
+                  *alloc_error = true;
+                  return return_code;
                 }
-
-                case SQL_C_WCHAR:
+                target_buffer = row[column_index].wchar_data;
+              }
+              else
+              {
+                row[column_index].char_data = (SQLCHAR *)malloc(buffer_size);
+                // if bad malloc, indicate and return
+                if (row[column_index].char_data == NULL)
                 {
-                  SQLWCHAR *memory_start = (SQLWCHAR *)data->bound_columns[column_index].buffer + (row_index * (data->columns[column_index]->buffer_size / sizeof(SQLWCHAR)));
-                  row[column_index].size = strlen16((const char16_t *)memory_start);
-                  row[column_index].wchar_data = new SQLWCHAR[row[column_index].size + 1]();
-                  memcpy
-                  (
-                    row[column_index].wchar_data,
-                    memory_start,
-                    row[column_index].size * sizeof(SQLWCHAR)
-                  );
-                  break;
+                  *alloc_error = true;
+                  return return_code;
                 }
+                target_buffer = row[column_index].char_data;
+              }
+
+              // Get the first chunk of data
+              return_code =
+              SQLGetData
+              (
+                data->hstmt,
+                column_index + 1,
+                data->columns[column_index]->bind_type,
+                target_buffer,
+                buffer_size,
+                &string_length_or_indicator
+              );
+              if (!SQL_SUCCEEDED(return_code))
+              {
+                return return_code;
+              }
```

Comment:
Shouldn't this also be wrapped in an `if (get_data_supports.block)` block?

Reply:
SQLGetData can still be called even if SQL_GD_BLOCK is false, it will just only work on a single row. As written, if a user specifies they want a cursor, set the fetch size greater than 1, and they have LONG data in their result set, they will get the long data from the first row over and over. Not sure the best way to handle it... some ideas:

1. If a user wants to fetch multiple rows of long data at once, and their driver doesn't support it, just fetch 1 at a time and stitch it together in the application to simulate a block fetch (none of the benefits, but the data is correct)
2. Have multiple branch paths, where each bullet is a new subpath:
  * User just wants to fetch all 1 at a time
  * User wants to fetch in blocks of X rows
    * User's driver can't do that, so just fetch 1 at a time and stitch together
    * User's driver can handle that for most data, but can't for long data using SQLGetData
      * Fetch 1 at a time and stitch together
      * Fetch SQLGetData with SQLBinCol
    * Fetch in X rows
3. Pre check if a user can block-fetch everything but LONG data, and if so, bind those with SQLBindCol how we used to

Reply:
I think to limit the scope of this PR, I will adjust the bind_col logic to follow option 3 for now.

Reply:
I definitely don't think that we should return the same row data over and over for long data. That's a giant trap that the user likely has no knowledge about (or shouldn't need to have).

I think option 1 makes the most sense. If the driver can't block fetch with SQLGetData, then fetch as many rows as the user asked for, just one at a time.

In any case, we should document these limitations. We might even want to provide an API (not in this PR, it's big enough as is) that users can call to get the capabilities in case they run in to issues - "The docs say if the driver doesn't support SQLGetData w/ block fetch it binds them which can cause issues for long data. How would I know that? Oh, call conn.checkDriverCapabilities() and check that 'supportsBlockFetchAndGetData' is true"

Reply:
An API to list capabilities sounds reasonable. For now I have implemented 3, just defaulting to SQLBindCol if the users driver cant handle multi-row SQLGetData. I will spend a little time thinking about the different combinations, but one thing that will _really_ help before trying to tackle something like that would be to get a robust multi-driver test suite running in GH Actions.